ogame/__init__.py

```diff
@@ -273,77 +273,106 @@ def collect_status(self):
             in_construction = False
         return is_possible, in_construction
 
+
+    def _parse_component(self, component, expected, cp):
+        '''
+        Find each "technology" and associated level or count
+
+        Used by supply(), facilities(), defense(), ships(), research()
+
+        In the future it would make more sense to return NamedTuples instead of custom classes
+        for each of those functions
+        '''
+        response = self.session.get(
+            url=self.index_php,
+            params={'page': 'ingame', 'component': component, 'cp': cp},
+        )
+        assert component in response.url, response.url
+        bs4 = self.BS4(response.text)
+
+        count_and_status = {}
+        for tech in bs4.find_all(class_='technology', attrs={'data-technology': True, 'data-status': True}):
+            tech_id = int(tech['data-technology'])
+            # buildings, faciliting, research use 'level'
+            # ships, defense use 'amount'
+            a = (tech.find(class_='amount', attrs={'data-value': True}) or
+                 tech.find(class_='level', attrs={'data-value': True}))
+            count_and_status[tech_id] = (int(a['data-value']), tech['data-status'])
+
+        found = set(count_and_status.keys())
+        expected = set(expected.values())
+        assert expected == found or (component == 'supplies' and expected.issubset(found)), (
+            "Didn't find all {} missing {}".format(component, expected - found))
+
+        return count_and_status
+
+
     def supply(self, id):
-        response = self.session.get(self.index_php + 'page=ingame&component=supplies&cp={}'.format(id)).text
-        bs4 = self.BS4(response)
-        levels = [int(level['data-value']) for level in bs4.find_all('span', {'data-value': True})]
-        status = [status['data-status'] for status in bs4.find_all('li', {'class': 'technology'})]
+        buildings = {name: const.buildings.__dict__[name][0]
+                for name in const.building_supply_names}
+        parsed = self._parse_component('supplies', buildings, id)
 
         class Supply:
-            def __init__(self, i):
-                self.level = levels[i]
-                self.data = OGame.collect_status(status[i])
+            def __init__(self, level, status):
+                self.level = level
+                self.data = OGame.collect_status(status)
                 self.is_possible = self.data[0]
                 self.in_construction = self.data[1]
 
         class Supplies(object):
-            metal_mine = Supply(0)
```

Comment:
Currently it's hard to fleetsave because it's difficult to convert the `class Ships` from `ships()` to a list of ships and counts.

This change adds a `count` field to each return class. Later it seems like these would benefit from being converted to NamedTuples at which point the 6 methods (ships, defense, research, buildings, facilities, moon_facilities) become one lines

```
def Supply():
     return self._parse_component('supplies', const.building_supply_names)

def Defense():
     return self._parse_component('defenses', const.building_defense_names)
```